tensorboard/uploader/uploader.py

```diff
@@ -541,19 +604,182 @@ def _create_point(self, tag_proto, event, value):
         return point
 
 
+class _BlobRequestSender(object):
+    """Uploader for blob-type event data.
+
+    Unlike the other types, this class does not accumulate events in batches;
+    every blob is sent individually and immediately.  Nonetheless we retain
+    the `add_event()`/`flush()` structure for symmetry.
+
+    This class is not threadsafe. Use external synchronization if calling its
+    methods concurrently.
+    """
+
+    def __init__(self, experiment_id, api, rpc_rate_limiter):
+        if experiment_id is None:
+            raise ValueError("experiment_id cannot be None")
+        self._experiment_id = experiment_id
+        self._api = api
+        self._rpc_rate_limiter = rpc_rate_limiter
+
+        self._new_request()
+
+    def _new_request(self):
+        """Declares the previous event complete."""
+        self._run_name = None
+        self._event = None
+        self._value = None
+        self._metadata = None
+
+    def add_event(
+        self, run_name, event, value, metadata,
+    ):
+        """Attempts to add the given event to the current request.
+
+        If the event cannot be added to the current request because the byte
+        budget is exhausted, the request is flushed, and the event is added
+        to the next request.
+        """
+        if self._value:
+            raise RuntimeError("Tried to send blob while another is pending")
+        self._run_name = run_name
+        self._event = event  # provides step and possibly plugin_name
+        self._value = value
+        # TODO(soergel): should we really unpack the tensor here, or ship
+        # it wholesale and unpack server side, or something else?
+        # TODO(soergel): can we extract the proto fields directly instead?
+        self._blobs = tensor_util.make_ndarray(self._value.tensor)
+        if np.ndim(self._blobs) != 1:
+            raise ValueError(
+                "A blob sequence must be represented as a rank-1 Tensor"
+            )
+        self._metadata = metadata
```

Comment:
AFAICT, `_blobs` isn’t initialized in either the initializer proper or
its `self._new_request()` call?

(Initializing in `_new_request` solves the confusing `AttributeError`
problem, of course, though I still think it’s a bit more readable to
explicitly initialize in the initializer, because our brains are trained
to scan that method for fields.)